service/history/queues/queue_immediate.go

```diff
@@ -0,0 +1,165 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queues
+
+import (
+	"sync/atomic"
+	"time"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/collection"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/quotas"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+var _ Queue = (*immediateQueue)(nil)
+
+type (
+	immediateQueue struct {
+		*queueBase
+
+		notifyCh chan struct{}
+	}
+)
+
+func NewImmediateQueue(
+	shard shard.Context,
+	category tasks.Category,
+	scheduler Scheduler,
+	executor Executor,
+	options *Options,
+	rateLimiter quotas.RateLimiter,
+	logger log.Logger,
+	metricsHandler metrics.MetricsHandler,
+) *immediateQueue {
+	paginationFnProvider := func(r Range) collection.PaginationFn[tasks.Task] {
+		return func(paginationToken []byte) ([]tasks.Task, []byte, error) {
+			ctx, cancel := newQueueIOContext()
+			defer cancel()
+
+			request := &persistence.GetHistoryTasksRequest{
+				ShardID:             shard.GetShardID(),
+				TaskCategory:        category,
+				InclusiveMinTaskKey: r.InclusiveMin,
+				ExclusiveMaxTaskKey: r.ExclusiveMax,
+				BatchSize:           options.BatchSize(),
+				NextPageToken:       paginationToken,
+			}
+
+			resp, err := shard.GetExecutionManager().GetHistoryTasks(ctx, request)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			return resp.Tasks, resp.NextPageToken, nil
+		}
+	}
+
+	return &immediateQueue{
+		queueBase: newQueueBase(
+			shard,
+			category,
+			paginationFnProvider,
+			scheduler,
+			executor,
+			options,
+			rateLimiter,
+			logger,
+			metricsHandler,
+		),
+
+		notifyCh: make(chan struct{}, 1),
+	}
+}
+
+func (p *immediateQueue) Start() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	p.logger.Info("", tag.LifeCycleStarting)
```

Comment:
add a component tag?

Reply:
Component tag will be added by the caller of NewImmdicate/ScheduleQueue.